kratos/spatial_containers/specialized_spatial_search.cpp

```diff
@@ -0,0 +1,353 @@
+    using PointType = PointObject<Element>;
+    using PointTypePointer = typename PointType::Pointer;
+    using PointVector = std::vector<PointTypePointer>;
+    using PointIterator = PointVector::iterator;
+    using DistanceIterator = std::vector<double>::iterator;
+
+    // Defining the PointVector
+    PointVector points = PrepareSearch(rStructureElements, rInputElements, rResults, rResultsDistance);
+
+    // Defining the search structure
+    if constexpr (TSearchBackend == SpatialContainer::KDTree) {
+        // Retrieving parameters
+        const int bucket_size = mParameters["bucket_size"].GetInt();
+
+        /// KDtree definitions
+        using BucketType = Bucket< 3ul, PointType, PointVector, PointTypePointer, PointIterator, DistanceIterator>;
+        using KDTree = Tree<KDTreePartition<BucketType>>;
+
+        // Creating the tree
+        KDTree kd_tree(points.begin(), points.end(), bucket_size);
+
+        // Performing search
+        ParallelSearch(rInputElements, rRadius, kd_tree, rResults, rResultsDistance);
+    } else if constexpr (TSearchBackend == SpatialContainer::Octree) {
+        // Retrieving parameters
+        const int bucket_size = mParameters["bucket_size"].GetInt();
+
+        /// Octree definitions
+        using BucketType = Bucket< 3ul, PointType, PointVector, PointTypePointer, PointIterator, DistanceIterator>;
+        using Octree = Tree<OCTreePartition<BucketType>>;
+
+        // Creating the tree
+        Octree octree(points.begin(), points.end(), bucket_size);
+
+        // Performing search
+        ParallelSearch(rInputElements, rRadius, octree, rResults, rResultsDistance);
+    } else if constexpr (TSearchBackend == SpatialContainer::BinsStatic) {
+        // Retrieving parameters
+        const int bucket_size = mParameters["bucket_size"].GetInt();
+
+        /// StaticBins definitions
+        using StaticBins = Bins<3ul, PointType, PointVector, PointTypePointer, PointIterator, DistanceIterator>;
+        using StaticBinsTree = Tree<StaticBins>;
+
+        // Creating the tree
+        StaticBinsTree static_bins_tree(points.begin(), points.end(), bucket_size);
+
+        // Performing search
+        ParallelSearch(rInputElements, rRadius, static_bins_tree, rResults, rResultsDistance);
+    } else if constexpr (TSearchBackend == SpatialContainer::BinsDynamic) {
+        /// BinsDynamic definitions
+        using DynamicBins = BinsDynamic<3ul, PointType, PointVector, PointTypePointer, PointIterator, DistanceIterator>;
+
+        // Creating the bins
+        DynamicBins dynamic_bins(points.begin(), points.end());
+
+        // Performing search
+        ParallelSearch(rInputElements, rRadius, dynamic_bins, rResults, rResultsDistance);
+    // } else if constexpr (TSearchBackend == SpatialContainer::BinsStaticObjects) {
+    // } else if constexpr (TSearchBackend == SpatialContainer::BinsDynamicObjects) {
+    } else {
+        KRATOS_ERROR << "Unknown search backend" << std::endl;
```

Comment:
the `else` branch here is a compile time error, so I'd use something like a `static_assert`

Reply:
Done